src/axom/primal/geometry/BezierPatch.hpp

```diff
@@ -0,0 +1,875 @@
+  }
+
+  /// Returns the order of the Bezier Patch on the first axis
+  int getOrder_u() const
+  {
+    return static_cast<int>(m_controlPoints.shape()[0]) - 1;
+  }
+
+  /// Returns the order of the Bezier Patch on the second axis
+  int getOrder_v() const
+  {
+    return static_cast<int>(m_controlPoints.shape()[1]) - 1;
+  }
+
+  /// Make trivially rational. If already rational, do nothing
+  void makeRational()
+  {
+    if(!isRational())
+    {
+      const int ord_u = getOrder_u();
+      const int ord_v = getOrder_v();
+
+      m_weights.resize(ord_u + 1, ord_v + 1);
+
+      for(int i = 0; i <= ord_u; ++i)
+        for(int j = 0; j <= ord_v; ++j) m_weights[i][j] = 1.0;
+    }
+  }
+
+  /// Make nonrational by shrinking array of weights
+  void makeNonrational() { m_weights.resize(0, 0); }
+
+  /// Use array size as flag for rationality
+  bool isRational() const { return (m_weights.size() != 0); }
+
+  /// Clears the list of control points, make nonrational
+  void clear()
+  {
+    const int ord_u = getOrder_u();
+    const int ord_v = getOrder_v();
+
+    for(int p = 0; p <= ord_u; ++p)
+      for(int q = 0; q <= ord_v; ++q) m_controlPoints(p, q) = PointType();
+
+    makeNonrational();
+  }
+
+  /// Retrieves the control point at index \a (idx_p, idx_q)
+  PointType& operator()(int ui, int vi) { return m_controlPoints(ui, vi); }
+
+  /// Retrieves the vector of control points at index \a idx
+  const PointType& operator()(int ui, int vi) const
+  {
+    return m_controlPoints(ui, vi);
+  }
+
+  /*!
+   * \brief Get a specific weight
+   *
+   * \param [in] ui The index of the weight on the first axis
+   * \param [in] vi The index of the weight on the second axis
+   * \pre Requires that the surface be rational
+   */
+  const T& getWeight(int ui, int vi) const
+  {
+    SLIC_ASSERT(isRational());
+    return m_weights(ui, vi);
+  }
+
+  /*!
+   * \brief Set the weight at a specific index
+   *
+   * \param [in] ui The index of the weight in on the first axis
+   * \param [in] vi The index of the weight in on the second axis
+   * \param [in] weight The updated value of the weight
+   * \pre Requires that the surface be rational
+   * \pre Requires that the weight be positive
+   */
+  void setWeight(int ui, int vi, T weight)
+  {
+    SLIC_ASSERT(isRational());
+    SLIC_ASSERT(weight > 0);
+
+    m_weights(ui, vi) = weight;
+  };
+
+  /// Checks equality of two Bezier Patches
+  friend inline bool operator==(const BezierPatch<T, NDIMS>& lhs,
+                                const BezierPatch<T, NDIMS>& rhs)
+  {
+    return (lhs.m_controlPoints == rhs.m_controlPoints) &&
+      (lhs.m_weights == rhs.m_weights);
+  }
+
+  friend inline bool operator!=(const BezierPatch<T, NDIMS>& lhs,
+                                const BezierPatch<T, NDIMS>& rhs)
+  {
+    return !(lhs == rhs);
+  }
+
+  /// Returns a copy of the Bezier patch's control points
+  CoordsMat getControlPoints() const { return m_controlPoints; }
```

Comment:
It might be unexpected for this to return a copy rather than a reference.
Should we rename the function to better indicate that it's a copy?

Note: We also want to be consistent w/ `primal::BezierCurve::getControlPoints()`, so if we change this one, we should change that one too. Perhaps we can reconsider in a follow-up PR.

Reply:
It might make more sense then to just change this method (and the one in `BezierCurve` to just return a reference as expected.

I'm working on another branch to fix minor things like this in `BezierCurve`, so regardless of whether I work this out in this branch or the next one, I'll make sure they're both consistent with one another at the same time.